File: BitManipulation/sum.cpp
#include <iostream>
#include <bitset>

class Solution {
public:
    int add(int a, int b) {
        while (b != 0) {
            int c = (unsigned int)(a & b) << 1;
            a ^= b;
            b = c;
            // std::bitset<10> bs(c);
            // std::cout << bs << '\n';
        }
        return a;
    }
};


int main()
{
    Solution sol;
    int res = sol.add(20, -7);
    std::cout << res << '\n';
    return 0;
}

File: BitManipulation/sum.py
class Solution:
    def add(self, a: int, b: int) -> int:
        mask = 0xffffffff
        a &= mask
        b &= mask
        while b:
            a, b = a ^ b, ((a & b) << 1) & mask
        # 0x7fffffff == (1 << 31) - 1
        return ~(a ^ mask) if a > 0x7fffffff else a


res = Solution().add(-1, -2)
print(res)